src/modules/navigator/mission_block.cpp

```diff
@@ -116,264 +124,301 @@ MissionBlock::is_mission_item_reached()
+		    dist_xy <= _navigator->get_acceptance_radius() &&
+		    dist_z <= _navigator->get_altitude_acceptance_radius()) {
 
-						if (next_sp.valid) {
-							_mission_item.yaw = get_bearing_to_next_waypoint(_navigator->get_global_position()->lat,
-									    _navigator->get_global_position()->lon,
-									    next_sp.lat, next_sp.lon);
+			success = true;
+		}
 
-							_waypoint_yaw_reached = false;
+	} else {
 
-						} else {
-							_waypoint_yaw_reached = true;
-						}
-					}
-				}
-			}
+		/* Loiter mission item on a non rotary wing: the aircraft is going to circle the
+		 * coordinates with a radius equal to the loiter_radius field. It is not flying
+		 * through the waypoint center.
+		 * Therefore the item is marked as reached once the system reaches the loiter
+		 * radius (+ some margin). Time inside and turn count is handled elsewhere.
+		 */
+		if (dist >= 0.0f &&
+		    dist_xy <= _navigator->get_acceptance_radius(fabsf(_mission_item.loiter_radius) * 1.2f) &&
+		    dist_z <= _navigator->get_altitude_acceptance_radius()) {
 
-		} else if (_mission_item.nav_cmd == NAV_CMD_DELAY) {
-			_waypoint_position_reached = true;
-			_waypoint_yaw_reached = true;
-			_time_wp_reached = now;
+			success = true;
+		}
+	}
 
-		} else {
-			/* for normal mission items used their acceptance radius */
-			float mission_acceptance_radius = _navigator->get_acceptance_radius(_mission_item.acceptance_radius);
+	return success;
+}
 
-			/* if set to zero use the default instead */
-			if (mission_acceptance_radius < NAV_EPSILON_POSITION) {
-				mission_acceptance_radius = _navigator->get_acceptance_radius();
-			}
+bool
+MissionBlock::yaw_achieved(const mission_item_s &item) const
+{
+	// return true if there's no yaw requirement
+	bool success = true;
 
-			/* for vtol back transition calculate acceptance radius based on time and ground speed */
-			if (_mission_item.vtol_back_transition) {
+	if (PX4_ISFINITE(item.yaw)) {
 
-				float velocity = sqrtf(_navigator->get_local_position()->vx * _navigator->get_local_position()->vx +
-						       _navigator->get_local_position()->vy * _navigator->get_local_position()->vy);
+		bool yaw_required = false;
 
-				const float back_trans_dec = _navigator->get_vtol_back_trans_deceleration();
-				const float reverse_delay = _navigator->get_vtol_reverse_delay();
+		if (_navigator->get_vstatus()->is_rotary_wing) {
+			// multicopter yaw setpoint for waypoint and loiter
+			if (item.nav_cmd == NAV_CMD_WAYPOINT ||
+			    item.nav_cmd == NAV_CMD_LOITER_UNLIMITED ||
+			    item.nav_cmd == NAV_CMD_LOITER_TIME_LIMIT) {
 
-				if (back_trans_dec > FLT_EPSILON && velocity > FLT_EPSILON) {
-					mission_acceptance_radius = ((velocity / back_trans_dec / 2) * velocity) + reverse_delay * velocity;
+				yaw_required = true;
 
-				}
+				const float yaw_err = _wrap_pi(item.yaw - _navigator->get_local_position()->yaw);
 
+				if (fabsf(yaw_err) > math::radians(_navigator->get_yaw_threshold())) {
+					success = false;
```

Comment:
The success and yaw_required boolean are a bit confusing to me. 
Should not first be checked that yaw is required, and if required (https://github.com/PX4/Firmware/pull/8883/files#diff-7a824e3d1e9a5420071c8c9bbe27b8e7R353) it then needs to be checked for success?